### ISSUE TITLE
Compatibility with LPAC

### ISSUE THREAD

**Issue by UVLabs** [fluidweb-co/fluid-checkout#47]
We started the conversation here: https://wordpress.org/support/topic/conflict-with-plugin-21/

Since then I was able to get the map to display by using the hook `fc_checkout_before_step_billing_fields` but two things:

1 - In your plugin billing fields come after shipping fields, so naturally I tried to find the hook `fc_checkout_before_step_shipping_fields` and it exists, but the map still gets stripped like it did with default WC hooks. Some users might want the map to the first thing on the page thats why i give them the option to change its orientation. 

2 - Every time the map is updated, a duplicate blank map area appears. I think this is because you said **_"An update happens right after the page is loaded, and when any significant field is changed."_** The thing is, every time a customer moves the marker on the map to their preferred location, the checkout fields are updated with the address details for that location...so how do would we stop this duplication issue? 

I zoomed out on the page so you can see an example of the duplication: https://prnt.sc/uUxtWhJJGU2d

This is the line where the map is being outputted to the checkout page, as you can see I've already added support for WooFunnels, would love to add full support for your plugin as well: https://github.com/UVLabs/location-picker-at-checkout-for-woocommerce/blob/master/includes/Bootstrap/Main.php#L221

**Comment by diegoversiani** [comment]
Hi @UVLabs,

> would love to add full support for your plugin as well

Sure, let's make it happen :)


1 - It turns out those hooks for the shipping address (`fc_checkout_before_step_shipping_fields` and `fc_checkout_after_step_shipping_fields`) are in the wrong place and the map is being replaced with the checkout fragments.

The correct behavior would be for these hooks to be outside the fragment so it does not get replaced with every update, in the same way as it already does with the related billing hooks.

I'll fix it and release an update. I let you know once the update is available.

2 - It seems that the duplicate map issue is related to the shipping section being replaced when it should not.

Could you confirm whether this happens also when using the hook `fc_checkout_before_step_billing_fields` (**for the billing address**)?

Best,
Diego

**Comment by diegoversiani** [comment]
Hi @UVLabs,

I made the necessary changes to fix both issues, which were caused by the same problem.

These changes will be available to users with the next update (1.5.8).

Could you check if everything works as expected with the beta version 1.5.8-beta-1?
https://t2449427.p.clickup-attachments.com/t2449427/93dab6a6-e42d-49fc-b655-cf0b24cca2bd/fluid-checkout-1.5.8-beta-1.zip

I let you know once the update is available to all users.

Best,
Diego

**Comment by UVLabs** [comment]
@diegoversiani  Looks like we're getting there. I've tested all 5 hooks and they seem to work as expected. But there still seems to be a bug with FC stripping out custom classes

In the Gif below, this field is supposed to be hidden by default, and this is how it's defined in my code:

```
$fields['billing']['lpac_save_address_checkbox'] = array(
			'type'     => 'checkbox',
			'label'    => $save_address_checkbox_label,
			'required' => false,
			'class'    => array( 'form-row-wide' ),
			'priority' => 1,
		);

$fields['billing']['lpac_saved_address_name'] = array(
			'type'     => 'text',
			'label'    => $enter_address_name_label,
			'required' => false,
			'class'    => array( 'form-row-wide', 'hidden' ),
			'priority' => 1,
		);
```

In JS I'm simply detecting when the user clicks the checkbox and remove the hidden class so that the field shows, then add it back if the checkbox is clicked again. But it seems FC is stripping out this custom 'hidden' before rendering:

![Apr-21-2022 18-33-49](https://user-images.githubusercontent.com/10324144/164562384-c30b0c7d-2b2d-4bc0-b2b7-aa6970147330.gif)

---

I also noticed another weird thing...this same field is duplicated in the billing area of FC for some reason:

<img width="603" alt="image" src="https://user-images.githubusercontent.com/10324144/164562986-59859e0f-5a65-4053-8af0-bc0e04ec9971.png">

FYI I'm adding these custom fields by filtering `woocommerce_checkout_fields` and adding my custom fields before returning the changed array.

**Comment by diegoversiani** [comment]
Hi @UVLabs,

Fluid Checkout calls the function `WC()->checkout()->get_checkout_fields()` a few times before rendending the page and/or fragments.

If you add the new field via the hook `woocommerce_checkout_fields` without checking if it is already added, it might be duplicated.

Another reason is that the field seems to be a `billing` field. Fluid Checkout will display all billing fields in the billing section. It seems that this field is only being considered a "billing only" field, which means they will be displayed after the checkbox "(billing) same as shipping address".

The saved address feature seems to be only available in the LPAC PRO version. For that reason, I'm not able to test this or check what else could be causing the issue.

Would you be willing to send me the LPAC PRO version (with a license valid for a few days) so I can check it on my side?

Best,
Diego

**Comment by UVLabs** [comment]
@diegoversiani the code for creating this field is simply:

```
function create_save_address_checkout_fields( $fields ) {

		$save_address_checkbox_label = __( 'Save the Shipping Address for the future?', 'map-location-picker-at-checkout-for-woocommerce' );
		$save_address_checkbox_label = apply_filters( 'lpac_save_address_for_future_text', $save_address_checkbox_label );

		$enter_address_name_label = __( 'Enter Address Name', 'map-location-picker-at-checkout-for-woocommerce' );
		$enter_address_name_label = apply_filters( 'lpac_enter_address_name_label', $enter_address_name_label, $fields );

		$fields['billing']['lpac_save_address_checkbox'] = array(
			'type'     => 'checkbox',
			'label'    => $save_address_checkbox_label,
			'required' => false,
			'class'    => array( 'form-row-wide' ),
			'priority' => 1,
		);

		$fields['billing']['lpac_saved_address_name'] = array(
			'type'     => 'text',
			'label'    => $enter_address_name_label,
			'required' => false,
			'class'    => array( 'form-row-wide', 'hidden' ),
			'priority' => 1,
		);

		return $fields;
}
apply_filters( 'woocommerce_checkout_fields', 'create_save_address_checkout_fields' );
```

In JavaScript I'm repositioning the field so that it appears below the map, thanks to your reply I was able to stop the initial duplication of the field, but I still need to prevent FC from rendering it again after the checkout is updated.

Here are the two problems which currently exist:

1) FC seems to strip the `hidden` class from the custom input field being created. I was able to fix this by adding it back in JS if it does not exist initially (which it should), but FC should not be removing classes added to custom checkout fields and should instead inherit them before outputting the field. 

2) When the checkout is updated, FC re-renders the checkout fields i.e adding the fields that we previously repositioned to the billing area again. As you mentioned, the solution seems to be to first check if the field already exists in the array before rendering it, but I think this needs to be done in FC, what filter exactly from your plugin can I use to prevent this? 

This seems to be a similar issue like before where the map area was being duplicated

You can send a message here:https://lpacwp.com/contact/ if you think you'd need to test on your end

**Comment by diegoversiani** [comment]
I see. The fields are added as normal fields and would not be duplicated in the `$fields` variable as they are always added with the same array key.

1. I agree. Fluid Checkout should not be removing the class `hidden`. I have made the changes and already tested what I could. For that reason, I'll indeed need the LPAC PRO version to see what is happening. I'll contact you.

2. Currently, there is no official way to prevent Fluid Checkout from outputting the fields when generating the checkout fragments.

Instead of adding the fields to the `$fields['billing']`, I suggest you output the fields at the final place they are shown by calling the function `woocommerce_form_field()` directly in the function that outputs the map placeholder element.

Alternatively, you could conditionally add the fields. In other words, only add the fields if the constant `DOING_AJAX` is not defined or not set to `true`. Although, I'm not sure it this could break something else.

**Comment by UVLabs** [comment]
Thanks for your comment @diegoversiani I have changed the way i defined the field to instead use `woocommerce_form_field()` and its working as expected now.

The first point regarding the hidden class doesn't exist anymore.

I have done some testing and everything seems to work fine now.

---
**Except** the places autocomplete feature of LPAC. I can see that you have a dedicated plugin for this but it's a free feature in my plugin...the places autocomplete script from google seems to be loading but the `focus` event doesn't seem to be getting set on the field which is what's needed by google.

I've tried equeuing my scrips after yours by setting `fc-bundles` as a dependency but that doesn't seem to make a difference.

Are you doing anything different to the checkout fields that might be preventing a 3rd party plugin from adding places autocomplete functionality to the `shipping_address_1` and `billing_address_1` ?

**Comment by diegoversiani** [comment]
Hi @UVLabs,

Most compatibility issues with Fluid Checkout revolve around the fact that it replaced fragments of the checkout page that the native WooCommerce checkout does not replace.

If you use the browser dev tools to block the checkout fragments request `*?wc-ajax=update_order_review` from loading, and therefore preventing the fields from being replaced, you'll see that the address suggestions start to work as expected.

In order to make the autocomplete feature work with my plugin Google Address Autocomplete, I re-initialize the components `google.maps.places.autocomplete` every time the checkout page fragments are reloaded.

You'll have to use `document.querySelector()` every time the checkout page fragments are updated in order to get the correct reference to the replaced fields in the DOM.

See below excerpt from my plugin code, with the irrelevant parts removed:

```js
        /**
	 * Initialize component and set related handlers.
	 * 
	 * @param   {Object}   options  Pass different settings values for initializing this component.
	 */
	_publicMethods.init = function( options ) {
		// [some code here]

		// Initialize address autocomplete fields
		initFields();
		if ( _hasJQuery ) {
			// Re-initialize address autocomplete fields after `updated_checkout`
			$( document.body ).on( 'updated_checkout', initFields );
		}
 		
		// [some more code here]
	};



        /**
	 * Initialize Google Address Autocomplete for all address lookup fields.
	 */
	var initFields = function() {
		var inputs = document.querySelectorAll( _settings.autocompleteInputSelector );
		inputs.forEach( initField );
	}



        /**
	 * Initialize Google Address Autocomplete for an address lookup field.
	 *
	 * @param   {HTMLElement}  input  Address lookup field element.
	 */
	var initField = function( input ) {
		// [ Define `autocompleteOptions` ]

		// Initialze Google Places Autocomplete
		var autocomplete = new google.maps.places.Autocomplete( input, autocompleteOptions );
		var onPlaceChange = function() {
			// [Handles the returned data and fill the address fields]
		}

		// Set event handler for suggestion selected/changed
		autocomplete.addListener( 'place_changed', onPlaceChange );

		// [some more code here]
	}
```

Best,
Diego

**Comment by UVLabs** [comment]
Ah, thank you @diegoversiani That seem to have done it. I'll add the remaining support for this feature in next update. With this i think its safe to say we've added support for each other's plugins. Thank you and good luck with sales 🤝

**Comment by diegoversiani** [comment]
Awesome @UVLabs! I have released the update 1.5.8 with the changes.
Thanks for the patience and good luck with your sales too :)
